include/sys/arch_interface.h

```diff
@@ -328,6 +328,29 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
  * @see ISR_DIRECT_DECLARE()
  */
 
+/**
+ * Arch-specific hook for allocating IRQs
+ *
+ * @return The newly allocated IRQ or UINT_MAX on error.
+ */
+unsigned int arch_irq_allocate(void);
```

Comment:
What's the difference between an IRQ that is "allocated" (which I assume means "has a x86 vector that won't be used by another driver"), "enabled" (can actually fire), "connected" (has a Zephyr handler registered) and "in use" (... honestly not sure?).

Right now we just have "connected" and "enabled".  If we're going to add more states to the model, we should be explicit about what they mean.  Isn't this something we could wrap together underneath irq_connect_dynamic()?

Reply:
This is really a VT-D requirements: being able to allocate/connect outside of arch_irq_connect_dynamic() 

arch_irq_connect_dynamic() is the default irq connector (and it allocates vector underneath): in case of VT-D being enabled, the irq/vector will be remapped (it will allocate an VT-D IRTE and won't call z_x86_irq_connect_on_vector. Once irq enablement lands on IOAPIC side: it will recore the allocated irte and program the IOAPIC accordingly). And that's something we need to avoid for VT-D owns interrupt. (it cannot/shouldn't remap it's own interrupt). 

I was thinking that perhaps adding a new parameter or at least a dedicated flag could have been a solution: but arch_irq_connect_dynamic() is meant to be generic and this VT-D stuff is really an x86 specific technology so it's a no-go.